### PR TITLE
Bump Iceberg and Nessie versions

### PR DESCRIPTION
**Pull request by ajantha-bhat**
## Description
- Bump Iceberg to 1.5.0 and Nessie to 0.77.1
- Remove deprecated class usage
- update ForwardingFileIo to handle https://github.com/apache/iceberg/pull/9592

## Additional context and related issues
https://iceberg.apache.org/releases/#150-release

## Release notes

(x) This is not user-visible or is docs only, and no release notes are required.

### CODE REVIEW

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/nessie/IcebergNessieTableOperations.java [hunk 1]
```diff
@@ -126,7 +128,11 @@ protected void commitToExistingTable(TableMetadata base, TableMetadata metadata)
     {
         verify(version.orElseThrow() >= 0, "commitToExistingTable called on a new table");
         try {
-            nessieClient.commitTable(base, metadata, writeNewMetadata(metadata, version.getAsInt() + 1), table, toKey(new SchemaTableName(database, this.tableName)));
+            if (table == null) {
+                table = nessieClient.table(toIdentifier(new SchemaTableName(database, tableName)));
```
Comment (nastra) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/nessie/IcebergNessieTableOperations.java:132]: it's not clear to me why these changes are necessary

Comment (ajantha-bhat) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/nessie/IcebergNessieTableOperations.java:132]: `nessieClient.commitTable` signature has changed. It just needs contentID instead of expected content. 

`ExpectedContent`(aka table)  was null before this PR also for `commitToExistingTable`. I have debug the original integration to understand it.

**Review by wendigo — APPROVED** [review]

**Comment by findinpath** [comment]
Do we need to update the docker image before or we'll do it in a follow-up (see https://github.com/trinodb/trino/pull/19564 ) ?

cc @wendigo

**Comment by findinpath** [comment]
> InputFile.length is being called since https://github.com/apache/iceberg/pull/9592

Pls provide relevant code link in the 1.5.0 tag of the Iceberg code.

**Comment by findinpath** [comment]
@wendigo could you please run the PR with secrets to see the output of the cloud tests as well?

**Comment by ebyhr** [comment]
/test-with-secrets sha=054aecc7acca690b80e048a7647c26924bb42fb8

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/nessie/IcebergNessieTableOperations.java [hunk 2]
```diff
@@ -126,7 +128,11 @@ protected void commitToExistingTable(TableMetadata base, TableMetadata metadata)
...
+            }
+            nessieClient.commitTable(base, metadata, writeNewMetadata(metadata, version.getAsInt() + 1), table.getId(),
+                    toKey(new SchemaTableName(database, this.tableName)));
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/nessie/IcebergNessieTableOperations.java:135]: `toKey()` method always takes AbstractIcebergTableOperations' `database` and `tableName` variables. We could simplify this method.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/nessie/IcebergNessieTableOperations.java (hunk 1 above)
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/nessie/IcebergNessieTableOperations.java:132]: I suppose this line doesn't throw NessieNotFoundException or NessieConflictException. I would move this `if` above `try`.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/nessie/IcebergNessieTableOperations.java [hunk 3]
```diff
@@ -109,7 +109,9 @@ protected void commitNewTable(TableMetadata metadata)
     {
         verify(version.isEmpty(), "commitNewTable called on a table which already exists");
         try {
-            nessieClient.commitTable(null, metadata, writeNewMetadata(metadata, 0), table, toKey(new SchemaTableName(database, this.tableName)));
+            String contentId = table == null ? null : table.getId();
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/nessie/IcebergNessieTableOperations.java:112]: nit: Move this line above `try`.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/nessie/IcebergNessieTableOperations.java [hunk 4]
```diff
@@ -144,8 +158,9 @@ protected void commitMaterializedViewRefresh(TableMetadata base, TableMetadata m
         throw new UnsupportedOperationException();
     }
 
-    private static ContentKey toKey(SchemaTableName tableName)
+    private static ContentKey toKey(String databaseName, String tableName)
     {
-        return ContentKey.of(Namespace.parse(tableName.getSchemaName()), tableName.getTableName());
+        SchemaTableName schemaTableName = new SchemaTableName(databaseName, tableName);
+        return ContentKey.of(Namespace.parse(schemaTableName.getSchemaName()), schemaTableName.getTableName());
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/nessie/IcebergNessieTableOperations.java:164]: Why not pass databaseName & tableName to `Namespace.parse()` directly?

Comment (ajantha-bhat) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/nessie/IcebergNessieTableOperations.java:164]: we miss the validation of empty schemaname and conversion of the identifiers into lower case.  
I don't know the background. But I assumed it was written for that purpose. So, retained the `SchemaTableName`

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergFileOperations.java [hunk 5]
```diff
@@ -197,6 +198,7 @@ public void testSelect()
                         .add(new FileOperation(SNAPSHOT, "InputFile.length"))
                         .add(new FileOperation(SNAPSHOT, "InputFile.newStream"))
                         .add(new FileOperation(MANIFEST, "InputFile.newStream"))
+                        .add(new FileOperation(MANIFEST, "InputFile.length"))
```
[the remark this answers is not in the document]
Comment (ajantha-bhat) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergFileOperations.java:201]: I think @rdblue can add a comment on this. 

I haven't explored this change in depth. @nastra has helped in fixing these test failures. 
Let me understand it aswell.

[the remark this answers is not in the document]
Comment (amogh-jahagirdar) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergFileOperations.java:201]: I think I see what's causing this change:
https://github.com/apache/iceberg/pull/9592/files#diff-2e0ffbcd66a9c894e677ff49fb760ccfb28191c7b7fa4e2dbe5f9b16d6a27560R184
we used to use the manifest.length() (the Iceberg maintained length), now we're using the[ object storage length ](https://github.com/apache/iceberg/pull/9592/files#diff-aeb2747061941360740730b65414fb2547bf560ce11762e02803e89dd76155a2R144)in tryCache. 

Not sure yet if that change is required for both encrypted and non-encrypted or just encrypted files.

[the remark this answers is not in the document]
Comment (amogh-jahagirdar) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergFileOperations.java:201]: Stepped through the debugger before/after the upgrade, it is related to that specific change I linked but doesn't really have anything to do with the tryCache part (that's only relevant for manifest caching but we don't exercise that path in these tests). 

Prior to the change we would do an inputFile.length() but we'd just used the already passed in manifest length from Iceberg metadata, and Trino's custom `TracingInputFile` used in these tests would correctly not count that as a length call since it's cached, and not hitting the file system. 

Now that we're not passing in the manifest length, we are actually making the object storage requests to derive the length and the Trino `TracingInputFile` will correctly log that request as a filesystem access.

[the remark this answers is not in the document]
Comment (amogh-jahagirdar) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergFileOperations.java:201]: Actually the problem is more that the default implementation of `FileIO#newInputFile(ManifestFile)` does not pass the length. 

We should fix the default implementation in Iceberg but we can avoid the regression in Trino by overriding `ForwardingFileIO#newInputFile(ManifestFile)` to have a correct implementation which passes the manifest length. @ajantha-bhat i'll raise a PR to your branch.

[the remark this answers is not in the document]
Comment (ajantha-bhat) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergFileOperations.java]: Thanks @amogh-jahagirdar. I have merged it. 

Do we need to open a discussion on Iceberg mailing list to properly fix it. Those who don't override will still go through an extra IO? Manually overriding for each fileIO is not efficient. DO we have to revert/update the Iceberg side change and have 1.5.1?

[the remark this answers is not in the document]
Comment (ajantha-bhat) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergFileOperations.java]: ping @danielcweeks, @rdblue

[the remark this answers is not in the document]
Comment (amogh-jahagirdar) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergFileOperations.java]: I've opened a PR on the Iceberg side to address this at the default implementation level https://github.com/apache/iceberg/pull/9953. We just need the fix in the default implementation, not a revert of the original commit.

Feel free to raise your thoughts on the dev list, but my thinking is that this regression isn't particularly significant to warrant a patch release but I can understand the concern, and it definitely doesn't feel great to have that in. 

I think it's not significant because the extra IO is a head request done specifically when reading manifests (the overriden cases for `DataFile` and `DeleteFile` are not exercised, but I put them there to be defensive). HEAD requests are (in the most extreme cases) upper 2 digit milliseconds, typically though 10-20ms. Reading manifests is typically a very small amount of time compared to the overall job/query time for most workloads, and typically only a small amount of manifests are read to begin with. 

All this leads me to believe this probably isn't worth addressing in a patch release, but I can certainly be convinced otherwise. Perhaps I'm generalizing too much about folk's workloads and how different query engines will behave!

**Comment by wendigo** [comment]
LGTM @findepi @findinpath do you want to chime in?

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergFileOperations.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (findepi) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergFileOperations.java]: Good find!

Can we get https://github.com/apache/iceberg/pull/9953 in?
Also, i think we should have static analysis for `ForwardingFileIo` -- https://github.com/trinodb/trino/pull/21079

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/fileio/ForwardingFileIo.java [hunk 6]
```diff
@@ -88,6 +92,36 @@ public void deleteFiles(Iterable<String> pathsToDelete)
...
+    }
+
+    @Override
+    public InputFile newInputFile(DeleteFile file)
+    {
+        checkArgument(
+                file.keyMetadata() == null,
+                "Cannot decrypt delete file: %s (use EncryptingFileIO)",
+                file.path());
+        return newInputFile(file.path().toString(), file.fileSizeInBytes());
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/fileio/ForwardingFileIo.java:122]: Does https://github.com/apache/iceberg/pull/9953 obsolete this?

at the minimum, let's have a TODO comment with a link

File: plugin/trino-iceberg/pom.xml [hunk 7]
```diff
@@ -239,12 +239,32 @@
...
             <version>${dep.nessie.version}</version>
+            <exclusions>
+                <exclusion>
+                    <groupId>com.fasterxml.jackson.core</groupId>
+                    <artifactId>jackson-annotations</artifactId>
+                </exclusion>
+                <exclusion>
+                    <groupId>com.fasterxml.jackson.core</groupId>
+                    <artifactId>jackson-databind</artifactId>
+                </exclusion>
```
Comment (findepi) [inline plugin/trino-iceberg/pom.xml:250]: Why these exclusions? document

is it safe to exclude these?

Comment (findepi) [reply plugin/trino-iceberg/pom.xml:250]: what is this needed for?

Comment (ajantha-bhat) [reply plugin/trino-iceberg/pom.xml]: Maven enforcer rule was giving a compile error when trino wasn't on the same jackson version. 
Looks like on latest Trino we are on the same version and this is not needed anymore. 

I have removed it now. Thanks for questioning it and making me to try again on the latest version.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/fileio/ForwardingFileIo.java (hunk 6 above)
Comment (ajantha-bhat) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/fileio/ForwardingFileIo.java:123]: added

**Review by findinpath — APPROVED** [review]

**Review by wendigo — APPROVED** [review]

**Comment by ajantha-bhat** [comment]
I don't remember mentioning Nessie version bumps in the release notes.

**Comment by ebyhr** [comment]
@mosabua No need to mention in the release note.
